### PR TITLE
Enable enclave exit when a TCS is blocked on I/O

### PR DESCRIPTION
**Pull request by mzohreva**
This fixes issue #109

### CODE REVIEW

**Comment by mzohreva** [comment]
Open questions:
- Do we need to add the exit handler for other calls such as `alloc`/`free`/`async_queues`/`insecure_time`, etc?
- Is there a better way to avoid creating an extra future?

File: enclave-runner/src/usercalls/mod.rs [hunk 1]
```diff
@@ -1083,6 +1092,23 @@ impl<T: UsercallExtension> From<T> for Box<dyn UsercallExtension> {
 struct UsercallExtensionDefault;
 impl UsercallExtension for UsercallExtensionDefault {}
 
+macro_rules! with_exit_handler {
```
Comment (jethrogb) [inline enclave-runner/src/usercalls/mod.rs]: This seems complicated, I'd much rather have something where we make use of tokio's task cancellation mechanisms such as https://docs.rs/tokio/0.2.13/tokio/runtime/struct.Runtime.html#method.shutdown_timeout

Comment (jethrogb) [reply enclave-runner/src/usercalls/mod.rs]: Also, if this is wrapping all usercalls, couldn't it be done one level down in the stack (in usercall dispatch)?

**Comment by jethrogb** [comment]
Can we add some tests? Maybe they could be part of the rust-lang/rust test suite.

File: enclave-runner/src/usercalls/mod.rs (hunk 1 above)
Comment (mzohreva) [reply enclave-runner/src/usercalls/mod.rs]: > use of tokio's task cancellation mechanisms such as https://docs.rs/tokio/0.2.13/tokio/runtime/struct.Runtime.html#method.shutdown_timeout

That would require upgrading to tokio 0.2, I'll explore that option using PR #210

> couldn't it be done one level down in the stack (in usercall dispatch)?

Good idea, I'll give that a shot

**Comment by mzohreva** [comment]
Just found out that the parallel future approach implemented here is not necessary. The blocking issue appears to be due to the following line in `syscall_loop()`:
```rust
tokio::runtime::current_thread::block_on_all(select_fut.unit_error().compat()).unwrap()
```
The documentation for `tokio::runtime::current_thread::block_on_all()` says:

> This first creates a new Runtime, and calls Runtime::block_on with the provided future, which blocks the current thread until the provided future completes. It then calls Runtime::run to wait for any other spawned futures to resolve.

The problematic part is the `It then calls Runtime::run to wait for any other spawned futures to resolve.` and that is not what we need in syscall_loop. We want to return from syscall_loop as soon as the select future returns.

I'll update this PR accordingly.

File: enclave-runner/src/usercalls/mod.rs [hunk 2]
```diff
@@ -776,14 +776,15 @@ impl EnclaveState {
         };
 
         let select_fut =
-            futures::future::select(return_future.boxed(), io_future.boxed()).map( |either| {
+            futures::future::select(return_future.boxed_local(), io_future.boxed_local()).map( |either| {
```
Comment (jethrogb) [inline enclave-runner/src/usercalls/mod.rs]: Please add a comment that you know `io_future` isn't going to resolve, but that's intentional.

**Pull request merged**